### ISSUE TITLE
Poor documentation

### ISSUE THREAD

**Issue by raul-dan** [constantcontact/php-sdk#135]
I'm in the process of integrating Constant Contact and I've spent the last 30 minutes trying to create a single contact using the API client - thats a lot considering the fact that the test is not even included in the app logic, it's just a CLI test.

I have to debug a lot of bad requests responses, I have no proper documentation to consult, I have to checkout the code every single time.

I recommend adding some basic use cases examples that would make the process straight forward. The documentation at http://developer.constantcontact.com/docs/contacts-api/contacts-index.html#create_contact is hard to follow.

**Comment by huebs** [comment]
I agree with this sentiment but disagree that better documentation is the best solution to this problem. A better solution would be a refactoring of the model classes to use setter/getter methods rather than publicly accessible properties. I've already done this in subclasses for parts of the library (which I'd be glad to convert into a pull-request). For example, here's my Contact subclass:

``` php
namespace App\ConstantContact\Components\Contacts;

use Ctct\Components\Contacts\Address;
use Ctct\Components\Contacts\Contact as BaseContact;
use Ctct\Components\Contacts\ContactList;
use Ctct\Components\Contacts\CustomField;
use Ctct\Components\Contacts\EmailAddress;
use Ctct\Components\Contacts\Note;
use DateTime;
use DateTimeZone;

/**
 * Adds accessor methods for class properties.
 *
 * @author Ben Huebscher <ben@hubtech.tv>
 */
class Contact extends BaseContact
{
    /**
     * @param BaseContact $contact
     */
    public function __construct(BaseContact $contact = null)
    {
        if (null !== $contact) {
            $this->id              = $contact->id;
            $this->status          = $contact->status;
            $this->first_name      = $contact->first_name;
            $this->last_name       = $contact->last_name;
            $this->confirmed       = $contact->confirmed;
            $this->source          = $contact->source;
            $this->email_addresses = $contact->email_addresses;
            $this->prefix_name     = $contact->prefix_name;
            $this->job_title       = $contact->job_title;
            $this->addresses       = $contact->addresses;
            $this->notes           = $contact->notes;
            $this->company_name    = $contact->company_name;
            $this->home_phone      = $contact->home_phone;
            $this->work_phone      = $contact->work_phone;
            $this->cell_phone      = $contact->cell_phone;
            $this->fax             = $contact->fax;
            $this->custom_fields   = $contact->custom_fields;
            $this->lists           = $contact->lists;
            $this->created_date    = $contact->created_date;
            $this->modified_date   = $contact->modified_date;
            $this->source_details  = $contact->source_details;
        }
    }

    /**
     * Enable backwards compatibility but emit a deprecation warning.
     *
     * @param $name
     *
     * @return mixed
     */
    public function __get($name)
    {
        @trigger_error('Directly accessing properties of '.__CLASS__.' is deprecated.', E_USER_DEPRECATED);

        return $this->$name;
    }

    /**
     * Enable backwards compatibility but emit a deprecation warning.
     *
     * @param $name
     * @param $value
     */
    public function __set($name, $value)
    {
        @trigger_error('Directly accessing properties of '.__CLASS__.' is deprecated.', E_USER_DEPRECATED);

        $this->$name = $value;
    }

    /**
     * @return bool
     */
    public function hasContactId()
    {
        return 0 !== $this->getContactId();
    }

    /**
     * @return int
     */
    public function getContactId()
    {
        return (int) $this->id;
    }

    /**
     * @return string
     */
    public function getStatus()
    {
        return $this->status;
    }

    /**
     * @return string
     */
    public function getFirstName()
    {
        return $this->first_name;
    }

    /**
     * @param string|null $firstName
     *
     * @return $this
     */
    public function setFirstName($firstName = null)
    {
        $this->first_name = null === $firstName ? null : (string) substr($firstName, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $firstName, $this->first_name);

        return $this;
    }

    /**
     * @return string
     */
    public function getLastName()
    {
        return $this->last_name;
    }

    /**
     * @param string|null $lastName
     *
     * @return $this
     */
    public function setLastName($lastName = null)
    {
        $this->last_name = null === $lastName ? null : (string) substr($lastName, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $lastName, $this->last_name);

        return $this;
    }

    /**
     * @return bool
     */
    public function isConfirmed()
    {
        return $this->confirmed;
    }

    /**
     * @return string
     */
    public function getSource()
    {
        return $this->source;
    }

    /**
     * @return null|string
     */
    public function getEmailAddress()
    {
        return 0 !== count($this->email_addresses) ? $this->email_addresses[0]->email_address : '';
    }

    /**
     * @param string|EmailAddress $emailAddress
     *
     * @return $this
     */
    public function setEmailAddress($emailAddress)
    {
        $this->email_addresses = [];

        return $this->addEmail($emailAddress);
    }

    /**
     * @param string|EmailAddress $emailAddress
     *
     * @return $this
     */
    public function addEmail($emailAddress)
    {
        if (!$emailAddress instanceof EmailAddress) {
            $emailAddress = new EmailAddress($emailAddress);
        }

        $emailAddress->email_address = null === $emailAddress->email_address ? null :
            (string) substr($emailAddress->email_address, 0, 80);

        self::triggerTruncationWarning(__METHOD__, $emailAddress, $emailAddress->email_address);

        $this->email_addresses[] = $emailAddress;

        return $this;
    }

    /**
     * @return string
     */
    public function getPrefixName()
    {
        return $this->prefix_name;
    }

    /**
     * @param string|null $prefixName
     *
     * @return $this
     */
    public function setPrefixName($prefixName = null)
    {
        $this->prefix_name = null === $prefixName ? null : (string) substr($prefixName, 0, 4);

        self::triggerTruncationWarning(__METHOD__, $prefixName, $this->prefix_name);

        return $this;
    }

    /**
     * @return string
     */
    public function getJobTitle()
    {
        return $this->job_title;
    }

    /**
     * @param string|null $jobTitle
     *
     * @return $this
     */
    public function setJobTitle($jobTitle = null)
    {
        $this->job_title = null === $jobTitle ? null : (string) substr($jobTitle, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $jobTitle, $this->job_title);

        return $this;
    }

    /**
     * @return Address[]
     */
    public function getAddresses()
    {
        return $this->addresses;
    }

    /**
     * @return $this
     */
    public function removeAddresses()
    {
        $this->addresses = [];

        return $this;
    }

    /**
     * @return Note[]
     */
    public function getNotes()
    {
        return $this->notes;
    }

    /**
     * @param Note[] $notes
     *
     * @return $this
     */
    public function setNotes($notes)
    {
        $this->notes = $notes;

        return $this;
    }

    /**
     * @return string
     */
    public function getCompanyName()
    {
        return $this->company_name;
    }

    /**
     * @param string|null $companyName
     *
     * @return $this
     */
    public function setCompanyName($companyName = null)
    {
        $this->company_name = null === $companyName ? null : (string) substr($companyName, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $companyName, $this->company_name);

        return $this;
    }

    /**
     * @return string
     */
    public function getHomePhone()
    {
        return $this->home_phone;
    }

    /**
     * @param string|null $homePhone
     *
     * @return $this
     */
    public function setHomePhone($homePhone = null)
    {
        $this->home_phone = null === $homePhone ? null : (string) substr($homePhone, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $homePhone, $this->home_phone);

        return $this;
    }

    /**
     * @return string
     */
    public function getWorkPhone()
    {
        return $this->work_phone;
    }

    /**
     * @param string|null $workPhone
     *
     * @return $this
     */
    public function setWorkPhone($workPhone = null)
    {
        $this->work_phone = null === $workPhone ? null : (string) substr($workPhone, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $workPhone, $this->work_phone);

        return $this;
    }

    /**
     * @return string
     */
    public function getCellPhone()
    {
        return $this->cell_phone;
    }

    /**
     * @param string|null $cellPhone
     *
     * @return $this
     */
    public function setCellPhone($cellPhone = null)
    {
        $this->cell_phone = null === $cellPhone ? null : (string) substr($cellPhone, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $cellPhone, $this->cell_phone);

        return $this;
    }

    /**
     * @return string
     */
    public function getFax()
    {
        return $this->fax;
    }

    /**
     * @param string|null $fax
     *
     * @return $this
     */
    public function setFax($fax = null)
    {
        $this->fax = null === $fax ? null : (string) substr($fax, 0, 50);

        self::triggerTruncationWarning(__METHOD__, $fax, $this->fax);

        return $this;
    }

    /**
     * @return CustomField[]
     */
    public function getCustomFields()
    {
        return $this->custom_fields;
    }

    /**
     * @param string $name
     * @param string $value
     *
     * @return $this
     */
    public function setCustomField($name, $value)
    {
        $customField = $this->getCustomField($name);

        if (null === $customField) {
            $this->custom_fields[] = $customField = new CustomField();

            $customField->name = $name;
        }

        $customField->value = $value;

        return $this;
    }

    /**
     * @param string $name
     *
     * @return CustomField|null
     */
    public function getCustomField($name)
    {
        foreach ($this->custom_fields as $customField) {
            if ($customField->name === $name) {
                return $customField;
            }
        }

        return null;
    }

    /**
     * @param string $name
     *
     * @return string
     */
    public function getCustomFieldValue($name)
    {
        foreach ($this->custom_fields as $customField) {
            if ($customField->name === $name) {
                return $customField->value;
            }
        }

        return '';
    }

    /**
     * @param int $contactListId
     *
     * @return bool
     */
    public function hasListId($contactListId)
    {
        foreach ($this->lists as $list) {
            if ("$contactListId" === "$list->id") {
                return true;
            }
        }

        return false;
    }

    /**
     * @param int $contactListId
     *
     * @return $this
     */
    public function addListId($contactListId)
    {
        if (!$this->hasListId($contactListId)) {
            $this->lists[] = new ContactList($contactListId);
        }

        return $this;
    }

    /**
     * @param int $contactListId
     *
     * @return $this
     */
    public function removeListId($contactListId)
    {
        $this->lists = array_values(
            array_filter(
                $this->lists,
                function (ContactList $list) use ($contactListId) {
                    return "$contactListId" !== "$list->id";
                }
            )
        );

        return $this;
    }

    /**
     * @return bool
     */
    public function hasLists()
    {
        return 0 !== count($this->lists);
    }

    /**
     * @return ContactList[]
     */
    public function getLists()
    {
        return $this->lists;
    }

    /**
     * @return DateTime|null
     */
    public function getCreatedDate()
    {
        return null === $this->created_date ? null :
            date_create($this->created_date)->setTimezone(new DateTimeZone(date_default_timezone_get()));
    }

    /**
     * @return DateTime|null
     */
    public function getModifiedDate()
    {
        return null === $this->modified_date ? null :
            date_create($this->modified_date)->setTimezone(new DateTimeZone(date_default_timezone_get()));
    }

    /**
     * @return string
     */
    public function getSourceDetails()
    {
        return $this->source_details;
    }

    /**
     * @param string $method
     * @param string $original
     * @param string $truncated
     */
    private static function triggerTruncationWarning($method, $original, $truncated)
    {
        if ($original !== $truncated) {
            @trigger_error("$method: $original was truncated to $truncated.", E_USER_WARNING);
        }
    }
}
```

**Comment by WoogieNoogie** [comment]
Just about every method is fully documented with PHPDoc and hosted at http://constantcontact.github.io/php-sdk/, the quickstart guide in the readme shows generally how to construct the ConstantContact object and execute requests, and there are 5 PHP files in the samples directory that are fully functional. I'm not sure what else we can really provide with this.

I agree, with you @huebs - having getters/setters instead of exposing the variables would be much better. We're kind of working off an older style of architecture, and most of the time I've spent has been updating the back end dependencies, and less on reformatting the old code styles. Feel free to submit any PR's, and I'll be happy to take a look at merging them in, but I'm not sure how long those changes would end up being around before we move to the newer schemas when V3 ships later this year.

**Comment by hozyali** [comment]
Hi, the example files are adding a file ie;  ../vendor/auto.php but there is no vendor when I downloaded the sdk. can you advise where to get the full sources. thanks

**Comment by huebs** [comment]
You need to run `composer install`.

**Comment by hozyali** [comment]
Thanks for a quick reply. Is there a way to get the source directly? Otherwise what would be the solution to integrate this service in a wordpress website via php ?

**Comment by huebs** [comment]
https://getcomposer.org

**Comment by shannon7wallace** [comment]
Hi Hozyali, Check the "standalone" folder in this repository; that's in the master branch. If you unzip the zip file contained there, it has all of the files you should need. I hope that helps.

**Comment by hozyali** [comment]
Got it. Thank you so much for helping out.
